Enforce the ProducesPerTokenLogProbabilities capability.

brianwgoldman opened the pull request:
The goal is to move capabilities from self reported to verified. Unfortunately it is still possible for a SUT to say it provides logprobs when it doesn't, due to the conditional nature of the request.

* Runtime assertion error if a SUT attempts to return `top_logprobs` without listing the capability.
* Tests will get a runtime assertion error if:
  * They set SUTOptions.top_logprobs without listing the capability
  * They list the capability but don't set SUTOptions.top_logprobs.

brianwgoldman commented on newhelm/sut_decorator.py:
```diff
@@ -43,3 +45,30 @@ def wrapped_init(self, *args, **kwargs):
     assert params[1].name == "uid", "All SUTs must have UID as the first parameter."
+
+
+def _override_translate_response(cls: Type[PromptResponseSUT]) -> None:
```
I'll add Tests if you all think this idea is worth the potential confusion it may cause.

bkorycki commented:
I wonder if it would be more direct to enforce a test's required capabilties via SUTOptions. Like if a test does not specify `ProducesPerTokenLogProbabilities`, then an error is thrown if `top_logprobs` is included in one of its test items' SUTOptions. If that is a specified requirement for the test, then `top_logprobs` defaults to 1.

That would also be helpful for a test maker who forgets to include `ProducesPerTokenLogProbabilities` as a requirement, and creates test items with `SUTOptions(top_logprobs=X)`. If they run their test with a SUT that does not have that capability, then there would be no exceptions that tell the user why the SUT is not returning logprobs.

brianwgoldman commented:
> I wonder if it would be more direct to enforce a test's required capabilties via SUTOptions.

I like it! I've replaced previous test_decorator behavior with asserts on SUTOptions, and added unit testing.
